plugins/thunderbolt/fu-thunderbolt-image-validation.c

```diff
@@ -0,0 +1,482 @@
+		{ .offset = 0x12,  .len = 1, .description = "PA", .mask = 0xCC, .section = DRAM_UCODE_SECTION },
+		{ .offset = 0x13,  .len = 1, .description = "PB", .mask = 0x44, .section = DRAM_UCODE_SECTION },
+		{ .offset = 0x121, .len = 1, .description = "Snk0" },
+		{ .offset = 0xB6,  .len = 1, .description = "PA/PB (USB2)", .mask = 0xC0 },
+		{ 0 }
+	};
+
+	switch (id)
+	{
+	case 0x156D:
+	case 0x156B:
+		return FR;
+	case 0x157E:
+		return WR;
+	case 0x1578:
+	case 0x1576:
+	case 0x15D3:
+	case 0x15DA:
+		return AR;
+	case 0x15C0:
+		return AR_LP;
+	default:
+		return NULL;
+	}
+}
+
+static FwLocation *
+get_device_locations (guint16 id) {
+	static const FwLocation locations[] = {
+		{ .offset = 0x124, .len = 1, .section = ARC_PARAMS_SECTION, .description = "X of N" },
+		{ 0 }
+	};
+
+	switch (id)
+	{
+	case 0x1578:
+	case 0x1576:
+	case 0x15D3:
+	case 0x15DA:
+	case 0x15C0:
+		return locations;
+	default:
+		return NULL;
+	}
+}
+
+/*
+ * Compares the given locations, assuming locations is an array.
+ * Returns FALSE and sets error upon failure.
+ * locations points to the end of the array (the empty entry) upon
+ * successful return.
+ */
+static gboolean
+compare_locations (const FwLocation **locations,
+		   const FwObject    *controller,
+		   const FwObject    *image,
+		   GError           **error)
+{
+	for ( ; (*locations)->offset != 0; (*locations)++)
+	{
+		if (! compare (*locations, controller, image)) {
+			g_set_error (error,
+				     FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+				     "FW image image not compatible to this controller (%s)",
+				     (*locations)->description);
+			return FALSE;
+		}
+	}
+	return TRUE;
+}
+
+gboolean
+fu_plugin_thunderbolt_validate_firmware_impl (GByteArray  *controller_fw,
+					      GBytes      *blob_fw,
+					      GError     **error)
+{
+	gboolean is_host;
+	const HwInfo *hw_info;
+	const FwLocation *locations;
+
+	gsize blob_size;
+	const guint8 *blob_data = g_bytes_get_data(blob_fw, &blob_size);
+
+	guint32 controller_sections[SECTION_COUNT] = { [DIGITAL_SECTION] = 0 };
+	guint32 image_sections     [SECTION_COUNT] = { 0 };
+
+	const FwObject controller = { controller_fw->data, controller_fw->len, controller_sections };
+	const FwObject image      = { blob_data,           blob_size,          image_sections };
+
+	const FwLocation is_host_loc   = { .offset = 0x10, .len = 1, .mask = 1 << 1 };
+	const FwLocation device_id_loc = { .offset = 0x5,  .len = 2 };
+
+	image_sections[DIGITAL_SECTION] = read_farb_pointer (&image);
+	if (! valid_farb_pointer(image_sections[DIGITAL_SECTION])) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+				     "Invalid FW image file format");
+		return FALSE;
+	}
+
+	is_host = read_bool (&is_host_loc, &controller);
+	hw_info = get_hw_info (read_uint16 (&device_id_loc, &controller));
+
+	if (hw_info == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED,
+			     "Unknown controller");
+		return FALSE;
```

Comment:
Rather than fully unsupported TBT updates when it's a new controller not yet in fwupd, what do you think about allowing force overrides that "skip" validation?

Reply:
I think allowing that when `FWUPD_INSTALL_FLAG_FORCE` is specified is a very good thing to have. In RHEL we can't update fwupd super quickly, and we want to support the latest hardware too.

Reply:
Yeah that's what I was particularly worried about, especially as TBT comes on chip in the future.

Reply:
Just this one before I'm offline for 24h:
With the CPU integrated TBT controller, the FW is expected to be part of the system bios image, so the update will be from fwupdate plugin, not from thunderbolt :)

For the support:
How the kernel supoort is added? Backporting kernel patches is easier than updating fwupd for RHEL?

Anyway, the force / skip validation is dangerous but it's up to the user to decide so I'm fine with adding it.

For devices, it may be possible to do even better and try comparing the vid/mid anyway. I'll think about it.

Reply:
> Backporting kernel patches is easier than updating fwupd for RHEL?

That's above my pay grade to say I think. Otherwise, do you want me to wait for this validation before I roll the release? I was hoping to roll the release before the weekend, but can wait a few days if you think the validation is very important before people try using the plugin.

Reply:
>With the CPU integrated TBT controller, the FW is expected to be part of the system bios image, so the update will be from fwupdate plugin, not from thunderbolt :)

True.  I didn't consider that.  

> Anyway, the force / skip validation is dangerous but it's up to the user to decide so I'm fine with adding it.

Sounds good.

Reply:
@hughsie I'm not sure if the two parts of your comment are related.
All this plugin makes no sense for RHEL if it doesn't pick also the new Thunderbolt driver from kernel v4.13.
The question is about the future. Being future-proof for supporting new HW helps only if the kernel patches to support this new HW can get in.

For validation of the code in this PR - as I said, it fails my tests. I'm working now on debugging it, mainly by comparing the results to the other (already working) implementation we have. I hope to fix it soon so it can be ready for merge and release.